akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.fusing
+
+import akka.stream._
+import akka.stream.stage._
+import akka.stream.scaladsl._
+import akka.stream.actor.ActorSubscriberMessage
+import akka.stream.actor.ActorSubscriberMessage._
+import akka.stream.actor.ActorPublisherMessage
+import akka.stream.actor.ActorPublisherMessage._
+import scala.concurrent.forkjoin.ThreadLocalRandom
+
+import java.{ util ⇒ ju }
+
+final class FlattenMerge[T, M](breadth: Int) extends GraphStage[FlowShape[Source[T, M], T]] {
+  private val in = Inlet[Source[T, M]]("flatten.in")
+  private val out = Outlet[T]("flatten.out")
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(attr: Attributes) = new GraphStageLogic(shape) {
+
+    var sources = Set.empty[LocalSource[T]]
+    def activeSources = sources.size
+
+    private trait Queue {
+      def hasData: Boolean
+      def enqueue(src: LocalSource[T]): Unit
+      def dequeue(): LocalSource[T]
+    }
+
+    private class FixedQueue extends Queue {
```

Comment:
Nice approach. I had a similar idea about having a circular linked list of arrays as a reusable dynamic, but cache friendly queue. Not for this ticket though, but it seems to be useful in many places to be investigated further.